module/zstd/zstd.c

```diff
@@ -0,0 +1,632 @@
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file and include the License file at usr/src/OPENSOLARIS.LICENSE.
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ */
+
+/*
+ * Copyright (c) 2016-2018, Klara Systems Inc. All rights reserved.
+ * Copyright (c) 2016-2018, Allan Jude. All rights reserved.
+ * Copyright (c) 2018-2019, Sebastian Gottschall. All rights reserved.
+ * Copyright (c) 2019, Michael Niewöhner. All rights reserved.
+ */
+
+#include <sys/param.h>
+#include <sys/sysmacros.h>
+#include <sys/zfs_context.h>
+#include <sys/zio_compress.h>
+#include <sys/spa.h>
+#include <sys/zstd/zstd.h>
+
+#define	ZSTD_STATIC_LINKING_ONLY
+#include <zstd.h>
+#include <zstd_errors.h>
+#include <error_private.h>
+
+/* for userspace compile, we disable error debugging */
+#ifndef _KERNEL
+#define	printk(fmt, ...)
+#endif
+
+/* User space tests compatibility */
+#ifndef _KERNEL
+#define	__init
+#define	__exit
+#endif
+
+/*
+ * These enums describe the allocator type specified by
+ * kmem_type in struct zstd_kmem.
+ */
+enum zstd_kmem_type {
+	ZSTD_KMEM_UNKNOWN = 0,
+	/* Allocation type using kmem_vmalloc */
+	ZSTD_KMEM_DEFAULT,
+	/* Pool based allocation using mempool_alloc */
+	ZSTD_KMEM_POOL,
+	/* Reserved fallback memory for decompression only */
+	ZSTD_KMEM_DCTX,
+	ZSTD_KMEM_COUNT,
+};
+
+/* Special structure for pooled memory objects */
+struct zstd_pool {
+	void *mem;
+	size_t size;
+	kmutex_t 		barrier;
+	time_t timeout;
+};
+
+/* Global structure for handling memory allocations */
+struct zstd_kmem {
+	enum zstd_kmem_type	kmem_type;
+	size_t			kmem_size;
+	struct zstd_pool	*pool;
+};
+
+/*
+ * Special fallback memory structure used for decompression only
+ * if we're running out of memory
+ */
+struct zstd_fallback_mem {
+	size_t			mem_size;
+	void			*mem;
+	kmutex_t 		barrier;
+};
+
+struct levelmap {
+	int32_t cookie;
+	enum zio_zstd_levels level;
+};
+
+/*
+ * ZSTD memory handlers
+ * For decompression we use a different handler which has the possiblity
+ * of a fallback memory allocation in case we run out of memory.
+ * We split up the zstd handlers for most simplified implementation.
+ */
+
+static void *zstd_alloc(void *opaque, size_t size);
+static void *zstd_dctx_alloc(void *opaque, size_t size);
+static void zstd_free(void *opaque, void *ptr);
+
+/* Compression memory handler */
+static const ZSTD_customMem zstd_malloc = {
+	zstd_alloc,
+	zstd_free,
+	NULL,
+};
+
+/* Decompression memory handler */
+static const ZSTD_customMem zstd_dctx_malloc = {
+	zstd_dctx_alloc,
+	zstd_free,
+	NULL,
+};
+
+/*
+ * Level map for converting zfs internal levels to zstd levels
+ * and vice versa
+ */
+static struct levelmap fastlevels[] = {
```

Comment:
You forgot to rename this struct. @c0d3z3r0

Reply:
thanks